### PR TITLE
make testing framework user-friendly and featureful

### PR DESCRIPTION
**Pull request by ignatenkobrain**
Key features:
* Python 2 and 3 compatible
* User-friendly step syntax
* Human-readable error messages
* Speed improvements
* Proper support for changes in rpmdb
  - without using dnf.Sack
  - by using RPM Python API
* Support for additional states in rpmdb
  - reinstalled
  - unchanged
* Checks that no packages appears/disappears during comparison
* Checks that all present packages are being unchanged
* Support for writing INI files
* Support for enabling/disabling repositories
* Documentation
* Auto-generated documentation

### CODE REVIEW

**Comment by ignatenkobrain** [comment]
Remaining thing is only to port existing tests.

File: dnf-docker-test/features/autoremove.feature [hunk 1]
```diff
@@ -1,36 +1,31 @@
...
-        | present    | bar       |
-      When I execute "dnf" command "-y autoremove" with "success"
-      Then transaction changes are as follows
-        | State      | Packages  |
-        | present    | foo       |
-        | removed    | bar       |
+  Background: Repositories "base" + "upgates" with dropped requires
+        Given repository "base" with packages
+           | Package | Tag      | Value |
+           | TestA   | Requires | TestB |
```
Comment (j-mracek) [inline dnf-docker-test/features/autoremove.feature]: version 1 and release 1 is as default?

Comment (ignatenkobrain) [inline dnf-docker-test/features/autoremove.feature]: yes.

File: Dockerfile [hunk 2]
```diff
@@ -2,8 +2,15 @@ FROM fedora:24
 ENV LANG C
 
 RUN echo "deltarpm=0" >> /etc/dnf/dnf.conf
+RUN curl https://copr.fedorainfracloud.org/coprs/ignatenkobrain/behave-enhanced/repo/fedora-24/ignatenkobrain-behave-enhanced-fedora-24.repo -o /etc/yum.repos.d/behave-enhanced.repo
```
Comment (j-mracek) [inline Dockerfile]: This is not good idea

File: dnf-docker-test/README.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Generate documentation
```
Comment (j-mracek) [inline dnf-docker-test/README.md]: Single readme file for whole project would be better

File: dnf-docker-test/features/steps/command_steps.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import command_utils
+
+@parse.with_pattern(r"stdout|stderr")
+def parse_stdout_stderr(text):
+    return text
+
+register_type(stdout_stderr=parse_stdout_stderr)
+
+@when('I run "{command}"')
+def step_i_run_command(ctx, command):
```
Comment (j-mracek) [inline dnf-docker-test/features/steps/command_steps.py]: Always  I would prefer command combine with output status.

File: dnf-docker-test/features/steps/command_steps.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    Run a ``{command}`` as subprocess, collect its output and returncode.
+    """
+    ctx.cmd_result = command_utils.run(ctx, command)
+
+@when('I successfully run "{command}"')
```
Comment (j-mracek) [inline dnf-docker-test/features/steps/command_steps.py]: duplicate - not need it

File: dnf-docker-test/features/steps/command_steps.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+def step_i_successfully_run_command(ctx, command):
+    step_i_run_command(ctx, command)
+    step_the_command_should_pass(ctx)
+
+@then("the command should pass")
```
Comment (j-mracek) [inline dnf-docker-test/features/steps/command_steps.py]: Not needed.

File: dnf-docker-test/features/steps/command_steps.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+def step_the_command_should_pass(ctx):
+    ctx.assertion.assertEqual(ctx.cmd_result.returncode, 0)
+
+@then("the command should fail")
```
Comment (j-mracek) [inline dnf-docker-test/features/steps/command_steps.py]: Redundant

File: dnf-docker-test/features/steps/command_steps.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+def step_the_command_should_fail(ctx):
+    ctx.assertion.assertNotEqual(ctx.cmd_result.returncode, 0)
+
+@then("the command {stream:stdout_stderr} should contain exactly")
+def step_the_command_stream_should_contain_exactly(ctx, stream):
+    ctx.assertion.assertIsNotNone(ctx.text, "Multiline text is not provided")
```
Comment (j-mracek) [inline dnf-docker-test/features/steps/command_steps.py]: The older part use contex instead of ctx. The code should use only one style.

File: dnf-docker-test/features/steps/command_utils.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+            # In PY2, shlex.split() requires bytes string (non-unicode).
+            # In PY3, shlex.split() accepts unicode string.
+            command = codecs.encode(command, "utf-8")
+        cmdargs = shlex.split(command)
+
+        command0 = cmdargs[0]
+        real_command = cls.COMMAND_MAP.get(command0, None)
+        if real_command:
+            cmdargs0 = real_command.split()
+            cmdargs = cmdargs0 + cmdargs[1:]
```
Comment (j-mracek) [inline dnf-docker-test/features/steps/command_utils.py]: No, it was previously rejected by jsilhan

File: dnf-docker-test/features/steps/command_utils.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+        try:
+            process = subprocess.Popen(cmdargs,
```
Comment (j-mracek) [inline dnf-docker-test/features/steps/command_utils.py]: The argument will be not extended by bash. Therefore re-designe it.

File: dnf-docker-test/features/steps/command_steps.py (hunk 5 above)
Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/command_steps.py]: not duplicate

File: dnf-docker-test/features/steps/command_steps.py (hunk 7 above)
Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/command_steps.py]: it's used for better readability to highlight that command should fail

File: dnf-docker-test/features/steps/command_steps.py (hunk 8 above)
Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/command_steps.py]: ctx is less to write and all my code use this style (hopefully one day we will remove old code)

File: dnf-docker-test/features/steps/command_utils.py (hunk 9 above)
Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/command_utils.py]: @jsilhan comments on this?

File: dnf-docker-test/features/steps/command_utils.py (hunk 10 above)
Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/command_utils.py]: We don't want to execute it by bash. If we want globbing works, I will implement it, but in a bit different way.

File: Dockerfile (hunk 2 above)
Comment (ignatenkobrain) [inline Dockerfile]: for a while it's good idea because it's still not pushed into repos. Once it will hit repos, I will change it once more.

File: dnf-docker-test/README.md (hunk 3 above)
Comment (ignatenkobrain) [inline dnf-docker-test/README.md]: I want to split project documentation (which is outdated) and specific things which are useful only to some people.

File: dnf-docker-test/features/steps/command_steps.py (hunk 4 above)
Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/command_steps.py]: in some cases for visibility it's better.

File: dnf-docker-test/features/steps/file_steps.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    """
+    table = Table(HEADINGS_INI)
+    def s2s(s):
+        return six.text_type(s, "utf-8") if six.PY2 else s
+    for section in configuration.sections():
+        for key, value in configuration.items(section):
+            table.add_row([s2s(section), s2s(key), s2s(value)])
+    return table
+
+@given('a file "{filepath}" with')
```
Comment (j-mracek) [inline dnf-docker-test/features/steps/file_steps.py]: no idea how to use it. The content of file is created by other line command?

File: dnf-docker-test/features/steps/file_steps.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+    ========= ===== =======
+     Section   Key   Value 
+    ========= ===== =======
+
+    Examples:
+
+    .. code-block:: gherkin
+
+       Feature: Creating INI files
```
Comment (j-mracek) [inline dnf-docker-test/features/steps/file_steps.py]: what would be an output?

File: dnf-docker-test/features/steps/file_steps.py (hunk 11 above)
Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/file_steps.py]: no, you provide multiline text, though I forgot to write example in documentation below

File: dnf-docker-test/features/steps/file_steps.py (hunk 12 above)
Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/file_steps.py]: ini file

File: dnf-docker-test/features/steps/repo_steps.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
+from __future__ import absolute_import
```
Comment (j-mracek) [inline dnf-docker-test/features/steps/repo_steps.py]: Split create repo staff into separate commit.

Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/repo_steps.py]: why?

Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/repo_steps.py]: I basically can't because it replaces already existing stuff.

File: dnf-docker-test/features/steps/command_utils.py (hunk 10 above)
Comment (j-mracek) [inline dnf-docker-test/features/steps/command_utils.py]: We want to mimic standard user environment and I think that for it is better to use shell=True

File: dnf-docker-test/features/steps/command_utils.py (hunk 9 above)
Comment (j-mracek) [inline dnf-docker-test/features/steps/command_utils.py]: Just see how many steps you perform before you run command by subproccess? Is it really worth to do them?

Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/command_utils.py]: yes, definitely. people should write HUMAN-READABLE/WRITEABLE text and not something what is easy to parse.

**Comment by j-mracek** [comment]
In general, please reduce testing statement to minimal number (1 is more than enough). Multiple expression of similar things doesn't help or add additional value when it can be replaced by one with additional variable.

**Comment by ignatenkobrain** [comment]
for each action there is only one statement.
